encoding/dot: Add decoding support for DOT graphs.

mewmew opened the pull request:
Updates gonum/graph#9.

mewmew commented:
Consider this a first pull request. I expect that there are issues we may wish to solve in a cleaner fashion (e.g. handling of subgraphs).

For now, I'd be very happy to get an intial feelback on the exposed API.

I know that this functionality is tremedously useful for my personal workflow with DOT graphs, and I'd love to see decoding functionality eventually merged with the gonum/graph project.

The included test case demonstrates how to create graphs with user-defined nodes and edges containing DOT attributes. The added test case exersises both encoding and decoding functionality, as it does a round-trip validation.

Cheers /u

mewmew commented:
I'd also be interested in discussion how to handle external dependencies.

[github.com/pkg/errors](https://github.com/pkg/errors) is likely to remain stable, as indicated by Dave Cheney.

[github.com/graphism/dot](https://github.com/graphism/dot) may either be vendored, copied or imported as an external dependency. The API of the [graphism/dot/ast](https://godoc.org/github.com/graphism/dot) is considered stable (see graphism/dot#6 for further details).

mewmew commented:
After playing around a bit with the API, I am considering moving the responsibility of calling `g.AddNode` and `g.SetEdge` from the user-defined `NewNode` and `NewEdge` (as is currently the case), to the DOT decoder.

The rationale is that this somewhat simplified the definition of `NewNode` and `NewEdge`, and more importantly, eleviates subtle bugs introduced if users forget to call `g.SetEdge` and `g.AddNode` themselves.

Still have to think a bit about it. If we may envision other use-cases for the newly defined methods `NewNode` and `NewEdge`, then it may still be worth keeping the definition of `NewNode` and `NewEdge` as is, putting the responsibility on the user to implement them correctly as documented. This would make it more pleasant to make use of `NewNode` and `NewEdge`, as they can be invoked and used directly.

Current example definition of `NewNode` and `NewEdge` for the user-defined node and edge types:

```go
// NewNode adds a new node with a unique node ID to the graph.
func (g *DirectedGraph) NewNode() graph.Node {
	return &Node{
		Node: simple.Node(g.NewNodeID()),
	}
}

// NewEdge adds a new edge from the source to the destination node to the graph,
// or returns the existing edge if already present.
func (g *DirectedGraph) NewEdge(from, to graph.Node) graph.Edge {
	if e := g.Edge(from, to); e != nil {
		return e
	}
	return &Edge{
		Edge: simple.Edge{F: from, T: to},
	}
}
```

Potential future definition:

```go
// NewNode adds a new node with a unique node ID to the graph.
func (g *DirectedGraph) NewNode() graph.Node {
	return &Node{
		Node: simple.Node(g.NewNodeID()),
	}
}

// NewEdge adds a new edge from the source to the destination node to the graph,
// or returns the existing edge if already present.
func (g *DirectedGraph) NewEdge(from, to graph.Node) graph.Edge {
	if e := g.Edge(from, to); e != nil {
		return e
	}
	return &Edge{
		Edge: simple.Edge{F: from, T: to},
	}
}
```

After writing down the potential future definition, one thing which may be very misleading with that definition is that `NewEdge` returns an edge that may or may not be part of the graph. In the previous definition of the API, `NewEdge` always returned an edge that was part of the graph.

Any thoughts?

mewmew commented:
Sorry for the long series of comments. Just wanted to mention, that this first pull request focues on the implementation of decoding. Once we have decided upon an API, it would be great to add native support for DOT decoding to the existing graph implementations (such as simple.DirectedGraph). Those implementations may be sent in follow-up pull requests, and should not be too difficult.

kortschak commented on encoding/dot/decode_test.go:
```diff
@@ -0,0 +1,210 @@
+		},
+	}
+	for _, g := range golden {
+		buf, err := ioutil.ReadFile(g.path)
```
This can be embedded and use a `strings.Reader`.

kortschak commented on encoding/dot/decode_test.go:
```diff
@@ -0,0 +1,210 @@
+package dot_test
```
Copyright header.

mewmew replied:
The reason dot_test was used, is because of collisions with the Node type, and to serve as an example of the package usage. However, none of those reasons truly justify it, so I'll rename the Node type to something else, and potentially add an example to `dot_example.go`.

kortschak commented on encoding/dot/decode.go:
```diff
@@ -0,0 +1,232 @@
+package dot
```
Copyright header.

kortschak commented on encoding/dot/decode_test.go:
```diff
@@ -0,0 +1,210 @@
+	"github.com/gonum/graph"
+	"github.com/gonum/graph/encoding/dot"
+	"github.com/gonum/graph/simple"
+	dotparser "github.com/graphism/dot"
```
If this stays, the graphism imports should be in their own block between std and gonum.

kortschak commented on encoding/dot/decode_test.go:
```diff
@@ -0,0 +1,210 @@
+	"github.com/gonum/graph/encoding/dot"
+	"github.com/gonum/graph/simple"
+	dotparser "github.com/graphism/dot"
+	"github.com/pkg/errors"
```
github.com/pkg/errors is not necessary here.

kortschak commented on encoding/dot/decode.go:
```diff
@@ -0,0 +1,232 @@
+					Value: attr.Val,
+				}
+				if err := n.UnmarshalDOTAttr(a); err != nil {
+					gen.errs = append(gen.errs, errors.Wrapf(err, "unable to unmarshal node DOT attribute (%v)", a))
```
Here I'd be inclined to have a specific error type that holds the kinds of information you want and has methods to retain the issues raised during the parse - something like os.PathError but with methods (possibly even a Cause method). I'm not convinced that retaining all errors is what we want though. You can just panic here, and recover the panic and return the error.

kortschak commented:
>I'd also be interested in discussion how to handle external dependencies.

We avoid them as much as possible. Just yesterday I moved one of my essentially static packages to gonum from a client package set to satisfy this. While deps are in theory not problematic, in practice where we have them, we have had problems with stability.

>github.com/pkg/errors is likely to remain stable, as indicated by Dave Cheney.

While the design principles of github.com/pkg/errors are sound, we don't need all the functionality that it provides (it is a generic case).

>github.com/graphism/dot may either be vendored, copied or imported as an external dependency. The API of the graphism/dot/ast is considered stable (see graphism/dot#6 for further details).

Since gonum is a library, we don't vendor. I personally would prefer to see either graphism dot be moved into the graph package suite or copied there.

mewmew commented:
> Since gonum is a library, we don't vendor. I personally would prefer to see either graphism dot be moved into the graph package suite or copied there.

Sounds good to me. We can move the graphism `dot` and `ast` packages into the gonum graph package suite. Should we place them in `encoding/dot` and `encoding/dot/ast`, respectively?

The primary reason for keeping graphism/dot a separate project is for users to be able to handle graph structures without being tied to the gonum/graph representation.

mewmew commented:
> While the design principles of github.com/pkg/errors are sound, we don't need all the functionality that it provides (it is a generic case).

Makes sense, I'll update the PR and remove the use of pkg/errors.

kortschak commented:
>Sounds good to me. We can move the graphism `dot` and `ast` packages into the gonum graph package suite. Should we place them in `encoding/dot` and `encoding/dot/ast`, respectively?

Yes, either that or in `encoding/dot/dotutils/...` or `encoding/formats/dot/...`. I think in `dot/...` as you have it should work well, but see how it goes.

>The primary reason for keeping graphism/dot a separate project is for users to be able to handle graph structures without being tied to the gonum/graph representation.

Understood. I had similar design principles in my bíogo packages hts and ncbi - they are separate from the broader bíogo packages and data types. I think the approach above still satisfies that.

mewmew commented:
> Yes, either that or in encoding/dot/dotutils/... or encoding/formats/dot/.... I think in dot/... as you have it should work well, but see how it goes.

May I transfer the `dot` repository to the gonum organization? That way the API may be kept separate from the gonum/graph representation while solving concerns with external dependencies.

kortschak commented:
We can do that, but it just delays the inevitable to decision of where to put it. We are in the slow process of deciding about merging repos.

mewmew commented:
> We can do that, but it just delays the inevitable to decision of where to put it. We are in the slow process of deciding about merging repos.

Ah, I see. Well, I'd be happy no matter where it ends up, as long as it may still be used as an isolated package if a user so wishes.

mewmew commented:
I'm heading off to bed. Thanks for the initial feedback and for starting the discussion on how we may get decoding support added to gonum/graph.

Good night Dan.

Cheers /u

kortschak commented:
>Ah, I see. Well, I'd be happy no matter where it ends up, as long as it may still be used as an isolated package if a user so wishes.

Maybe then in "encoding/formats/dot" then. That gives a clear enough indication that it is separate.

mewmew commented:
> Maybe then in "encoding/formats/dot" then. That gives a clear enough indication that it is separate.

Sounds good to me. I'll update the PR to get a feel for the layout of packages. We may consider changing the name of the encoding/formats/dot package to something else, so that users may import both dot packages without locally renaming one of them. Perhaps dotparser or something. We can bikeshed about different ideas for a bit : ) I'll update the PR meanwhile.

mewmew commented:
How about encoding/dot/dotparser? I'll stop with the bikesheding for now.

mewmew commented:
The PR should be ready for another round of reviews. For now, I've added the DOT parser to encoding/dot/dotparser, and updated all import paths accordingly.

The error handling has been simplified and gonum copyright notices have been added throughout.

Any additional things we may wish to polish or clean up before merging?

I do think there are cleaner way to handle subgraph decoding, but for now, the implementation works. And we may easily change the implementation in the future, without affecting the exported API.

mewmew commented:
@kortschak Would it be possible to specify somehow that the DOT parser code is also available under a public domain license? Several projects provide code under dual licenses, such as MIT and public domain. E.g.

https://github.com/nothings/stb#whats-the-license

mewmew commented:
Adding a dual license to the code would be as simple as adding a README to the encoding/dot/dotparser directory stating that the code is available under a dual license. Something along the lines:

# dotparser

## License

The source code and any original content of the dotparser directory is released into the [public domain](https://creativecommons.org/publicdomain/zero/1.0/).

The source code is also licensed under a BSD open source license, and users are free to choice the license which suits their needs.

Please see github.com/gonum/license for general license information, contributors, authors, etc on the Gonum suite of packages.

kortschak commented:
I'm OK with releasing under a dual license with CC0, though others should also agree to this. I would feel more comfortable if it is license under CC0 *by you* and licensed under the gonum license by the gonum AUTHORS.

I will get to the review proper later today.

/cc @gonum/developers

kortschak commented on encoding/dot/dotparser/ast/ast_test.go:
```diff
@@ -0,0 +1,95 @@
+
+// Verify that all statements implement the Stmt interface.
+var (
+	_ ast.Stmt = &ast.NodeStmt{}
```
Conventionally we do `_ ast.Stmt = (*ast.NodeStmt)(nil)`.

mewmew replied:
Sure. Will update the code.

kortschak commented on encoding/dot/dotparser/internal/astx/astx.go:
```diff
@@ -0,0 +1,322 @@
+// === [ File ] ================================================================
+
+// NewFile returns a new file based on the given graph.
+func NewFile(graph interface{}) (*ast.File, error) {
```
Are you expecting that this set of acceptable types will grow? (and below) `interface{}` tends to make me queasy unless strictly necessary.

mewmew replied:
Not at all, the accepted type of `NewFile` will always be `*ast.Graph`. See [dot.bnf](https://raw.githubusercontent.com/mewpull/graph/83ade5cff70af2c164be0fe012f5b47cf65244e1/encoding/dot/dotparser/internal/dot.bnf) for the DOT grammar read by Gocc.

The Gocc parser generator is invoking `astx.NewFile`, when a *reduce* action is triggered for the `File` production rule, which only includes the `Graph` non-terminal; which has the type of the return value from `astx.NewGraph`.

I realize that the `interface{}` arguments makes you feel queasy, I would tend to agree. Unfortunately, this is by design of Gocc, to make it possible for users to write BNF grammars with unknown reduce actions. If the reduce actions were known to Gocc in advance, the reduce function arguments may have more narrow types.

Truely, I don't know if there is a clean way to improve this situation using parser generators in Go. If there is, I would be most interested in discussing how, so we may improve the situation for all Gocc users.

That is the main reason why the Gocc specific and generated packages are put inside an `internal` directory. The API is not intended to be called by anyone, except Gocc.

kortschak commented on encoding/dot/dotparser/testdata/.gitignore:
```diff
@@ -0,0 +1,5 @@
+*.dot
```
Can you remove this file please.

mewmew replied:
There is a reason for putting it there, specifically in the `testdata` directory. And that is that the Makefile downloads the entire test suite of Graphviz, and converts a huge number of DOT files, tracking differences between the Graphviz implementation and the Go DOT parser. The intention is not to include original Graphviz test cases into this repo, but rather to use `make fetch` to download them.

mewmew commented on encoding/dot/decode.go:
```diff
@@ -0,0 +1,223 @@
+
+// Copy copies the nodes and edges from the Graphviz AST source graph to the
+// destination graph. Edge direction is maintained if present.
+func Copy(dst Builder, src *ast.Graph) (err error) {
```
Makes sense. The main problem is that a DOT file may contain more than one graph. For this reason, it is up to the user to parse a DOT file, and include *the* graph they are interested in, which most of the time is the only graph.

kortschak replied:
Can I suggest then that you have `Unmarshal(src []byte, v interface{}) error` (or `src io.Reader`) and then type switch on `v` so that if it is `*[]dot.Builder` you unmarshal all the graphs in `src` into the elements of `v`, and if it is `dot.Builder` you unmarshal the single graph in `src` into it and return an error if there is more that one graph described by `src`. The use of a `[]dot.Builder` is a bit unpleasant here, but I expect it would be the minority case.

mewmew replied:
I agree that using something along the lines of `Unmarshal(src []byte, v interface{}) error` would be cleaner, as then, the DOT AST details would not have to be explosed to the user.

I tried to implement this, but got stuck on a stumbling block.

It works well for when `v` is of type `dot.Builder`. However, using `*[]dot.Builder` doesn't seem to work, as the slice provided by the user would be empty, and the `Unmarshal` function would then append the decoded graphs to the slice. This would work, except, the `Unmarshal` function needs a user provided graph builder to be able to invoke the `NewNode` and `NewEdge` functions. Otherwise there is no way for the `Unmarshal` function to create user-defined types of nodes and edges.

Ideally. the user would know how many graphs are contained within a given file, and may then pass that number of `dot.Builder`s in a slice. If this was the case, the type of `v` could be `[]dot.Builder` instead of `*[]dot.Builder`, as the `Unmarshal` function would never append any new builders by itself, and thus does not need a pointer to slice.

Another approach would be for the user to pass one builder in the `*[]dot.Builder` slice and introduce some sort of `Dup` method on builders, to allow `Unmarshal` to create the number of builders needed from a single user-defined graph builder. Personally, I don't like this approach, as it makes the interface of dot.Builder even larger, and the implementation of `Dup` may at times be non-trivial depending on the representation of the user-defined graph.

Any ideas? What am I missing. I really hope we may resolve this final issue to get a clean API to work with.

kortschak replied:
One possible idea is that we make the encoding/dot functionality a subset of the language proper and only allow single graphs in the src. I am not at all strongly opposed to this; if people want the complete case they can deal with formats/dot.

mewmew replied:
> One possible idea is that we make the encoding/dot functionality a subset of the language proper and only allow single graphs in the src. I am not at all strongly opposed to this; if people want the complete case they can deal with formats/dot.

Sounds good to me! I'll update the API to something along the lines of:
```go
// Unmarshal parses the Graphviz DOT-encoded data and stores the result in dst.
// For DOT-files containing more than one graph, use dot.Copy.
func Unmarshal(data []byte, dst Builder) error {
}
```

mewmew commented on encoding/dot/decode.go:
```diff
@@ -0,0 +1,223 @@
+					Key:   attr.Key,
+					Value: attr.Val,
+				}
+				if err := n.UnmarshalDOTAttr(a); err != nil {
```
I also felt this before, and wrote my code as:

```go
err := foo()
if err != nil {}
```

before, rather than.

```go
if err := foo() err != nil {}
```

However, as with may things in Go, I tried to learn from the standard library, and while initially I did not feel it was right, I learnt to appreciate it. Same with the placement of braces. Coming from C, I always placed them on their own line. Now having been using Go with gofmt for such a long time, I even started writing C code as

```c
int f() {
}
```

rather than

```c
int f()
{
}
```

The reason for placing return values in the *init* statement of if-statements, is to limit their scope.

Taking a quick look at the Go standard library, it seems that "if err := foo(); err != nil {}" is more common than "err := foo()"

```bash
[~/go/src]$ grep -R -P "\tif err :=" . | wc -l
1323
[~/go/src]$ grep -R -P "\terr :=" . | wc -l
524
```

That being said, I don't care too much, and if the prefered style of gonum is to keep the condition on a separate line, I'd can easily update the PR. Just wanted to give some background.

kortschak replied:
Ack.

mewmew commented on encoding/dot/decode.go:
```diff
@@ -0,0 +1,223 @@
+}
+
+// unique returns the set of unique nodes contained within ns.
+func unique(ns []graph.Node) []graph.Node {
```
I'll take a look at intsets.

mewmew commented on encoding/dot/decode_test.go:
```diff
@@ -0,0 +1,225 @@
+// newUndirectedGraph returns a new undirected capable of creating user-defined
+// nodes and edges.
+func newUndirectedGraph() *UndirectedGraph {
+	return &UndirectedGraph{
```
Sure.

mewmew commented on encoding/dot/decode_test.go:
```diff
@@ -0,0 +1,225 @@
+	default:
+		return fmt.Errorf("unable to unmarshal node DOT attribute with key %q", attr.Key)
+	}
+	return nil
```
The reason I used a switch is because the test case was also meant to serve as an example, and then it would be easy for a user to extend the supported DOT attributes. I can update the code to use an if-statement instead, and use a switch in `dot_example.go` with two attributes instead of one.

mewmew commented on encoding/dot/decode_test.go:
```diff
@@ -0,0 +1,225 @@
+
+// DOTAttributes returns the DOT attributes of the node.
+func (n *Node) DOTAttributes() []dot.Attribute {
+	var attrs []dot.Attribute
```
Same rationale here, if users wanted to add more than one attribute, it is easy to extend. I'll update the code of the test case. In the example I'll add another attribute and keep the current style.

mewmew commented:
> On thinking more about the placement of the low level code, I think that it probably should go in "github.com/gonum/graph/formats/dot/...". This exposes it more to non-gonum users and avoids the stutter in the package use. For gonum users there is a low probability of package name collision since most won't be using the parser/ast. Including in ".../formats/" opens to door for other parsers to go there too.

I agree that placing the DOT parser in `gonum/graph/formats/dot` is a better choice. It avoids stuttering, and limits the risk of confusion of coupling with the graph.Graph representation.

The only issue is that with the current API, users would have to import the formats/dot package, as it is the only one exposing dot file parsing functions.

I would be very happy to solve this issue by considering alternative approaches to the `dot.Copy` function. As you already pointed out, a clean API would be dot.Unmarshal(r, dst), and that would effectively solve the issue, as `encoding/dot` would import `formats/dot` under the scenes and the user would normally not have to import `formats/dot` at all.

However, as previously mentioned, we do have an issue when DOT files contain more than one graph. Another potential issue is if the DOT file contains zero graphs. Should we treat both those cases as an error? Zero cases should be an error, but supporting files with multiple DOT graphs would be good somehow.

I'll move the dotparser to formats/dot, as it is indeed a better place. We still need to figure out the API of `encoding/dot.Unmarshal`.

Ideas for how the API may look like?

kortschak commented:
>The only issue is that with the current API, users would have to import the formats/dot package, as it is the only one exposing dot file parsing functions.

Understood. I think the suggestion above may help with that. Thanks.

kortschak commented on encoding/dot/decode_test.go:
```diff
@@ -84,142 +83,124 @@ const undirected = `graph H {
 // containing DOT attributes.
 
-// DirectedGraph extends simple.DirectedGraph to add NewNode and NewEdge
+// dotDirectedGraph extends simple.DirectedGraph to add NewNode and NewEdge
```
In test code, I usually just call the implementations things like `directedGraph`, `node` and `edge`. No need to change these though.

mewmew commented:
> I'm OK with releasing under a dual license with CC0, though others should also agree to this. I would feel more comfortable if it is license under CC0 by you and licensed under the gonum license by the gonum AUTHORS.

Glad to hear. I'd love to have it under a dual-license, so that those who wish to use it as public domain are free to do so. Does the readme need any updates, or may I add it to the formats/dot directory?

# formats/dot

## License

The source code and any original content of the formats/dot directory is released into the [public domain](https://creativecommons.org/publicdomain/zero/1.0/).

The source code is also licensed under a BSD open source license, and users are free to choice the license which suits their needs.

Please see github.com/gonum/license for general license information, contributors, authors, etc on the Gonum suite of packages.

mewmew commented:
I'll add an initial version of the readme, and we can have the discussion related to revisions of the file.

kortschak commented on formats/dot/README.md:
```diff
@@ -0,0 +1,9 @@
+
+## License
+
+The source code and any original content of the formats/dot directory is released into the [public domain](https://creativecommons.org/publicdomain/zero/1.0/).
```
I think this is not quite right. While the effect is like the intention of releasing the code into the public domain, it is actually a non-obliging license, I would say something like "The source code and any original content of the formats/dot directory is released under [Public Domain Dedication](https://creativecommons.org/publicdomain/zero/1.0/)." This includes the notion that is intended but avoids possible confusion with actual public domain, which may not be possible in some places.

mewmew replied:
Thanks, will update.

kortschak commented on formats/dot/README.md:
```diff
@@ -0,0 +1,9 @@
+
+The source code and any original content of the formats/dot directory is released into the [public domain](https://creativecommons.org/publicdomain/zero/1.0/).
+
+The source code is also licensed under a BSD open source license, and users are free to choice the license which suits their needs.
```
s/a BSD open source license/the gonum license/

mewmew replied:
I'll include an update.

mewmew commented:
There we go, the API has been updated. Users with DOT-files containing multiple graphs are referred to use dot.Copy, while dot.Marshal should cover most every day uses.

I feel very happy with how the API turned out.

Thanks a lot for the review and feedback Dan!

mewmew commented:
Unless I've missed something, the latest version of the PR should have addressed the feedback you gave.

It is a little hard using the GitHub web interface to know for certain. I wish there was a list bullet list with your comments, that had a check box for each comment which one may use to keep track of changes to make.

Let me know if there still is something in this first round that needs to be fixed. We can always add improvements later on.

kortschak commented on encoding/dot/decode_test.go:
```diff
@@ -61,7 +51,7 @@ func TestRoundTrip(t *testing.T) {
 }
 
-const directed = `digraph G {
+const directed = `digraph {
```
Does the `Unmarshal` call still work with the graph label present?

mewmew replied:
> Does the `Unmarshal` call still work with the graph label present?

It still works, but the graph ID information is lost. We could implement an `UnmarshalDOTID` or something along those lines, so that the ID may be unmarshalled, and later marshalled by graphs implementing the `DOTID` method.

Note, there are several other pieces of information in DOT files that is currently lost, and I don't know if the graph/encoding/dot package should aim for full coverage; the API would end up looking rather horrible I reckon.

For more involved decoding needs, users may import formats/dot directly.

I am definitely open to suggestions, both on how the API may look like and where to set the bar. Perhaps it makes sense to include unmarshalling support for graph IDs in dot.Builder, but ignore information such as the strictness of a graph, or other very Graphviz DOT-specific details.

You may sway my mind, with convincing arguments though :)

kortschak replied:
We can add supporting methods at a later stage if people ask for them. However, I think the labels here should be put back, purely to demonstrate that your code here does handle (though lossily) DOT with labels.

mewmew commented:
> This looks good. I'll spend hopefully some more time over the weekend doing a final once over and give someone else a chance to look.
>
> Thank you for bearing with me and putting in this work.

Sounds great! It has been a fun ride.

I came upon a realisation in my dreams last night, that we may safely unexport the `Copy` function without losing any functionality for more involved use-cases.

Example parsing of DOT-file containing multiple graphs, using only `dot.Unmarshal` follows.

```go
const multi = `digraph {
	// Node definitions.
	0 [label="foo 2"];
	1 [label="bar 2"];

	// Edge definitions.
	0 -> 1 [label="baz 2"];
}

graph {
	// Node definitions.
	0 [label="foo 2"];
	1 [label="bar 2"];

	// Edge definitions.
	0 -- 1 [label="baz 2"];
}`

func ExampleParseMultiple() {
	file, err := dotparser.ParseString(multi)
	if err != nil {
		log.Fatal(err)
	}
	for _, g := range file.Graphs {
		var dst dot.Builder
		if g.Directed {
			dst = newDotDirectedGraph()
		} else {
			dst = newDotUndirectedGraph()
		}
		data := []byte(g.String())
		if err := dot.Unmarshal(data, dst); err != nil {
			log.Fatal(err)
		}
		buf, err := dot.Marshal(dst, "", "", "\t", false)
		if err != nil {
			log.Fatal(err)
		}
		fmt.Println(string(buf))
		// Output:
		// digraph {
		// 	// Node definitions.
		// 	0 [label="foo 2"];
		// 	1 [label="bar 2"];
		//
		// 	// Edge definitions.
		// 	0 -> 1 [label="baz 2"];
		// }
		// graph {
		// 	// Node definitions.
		// 	0 [label="foo 2"];
		// 	1 [label="bar 2"];
		//
		// 	// Edge definitions.
		// 	0 -- 1 [label="baz 2"];
		// }
	}
}
```

Based on this insight, I'll unexport the `Copy` function and thus remove the user visible details of the formats/dot/ast package from the encoding/dot package, which feels like a much cleaner abstraction.

kortschak commented on encoding/dot/decode.go:
```diff
@@ -41,12 +41,12 @@ func Unmarshal(data []byte, dst Builder) error {
-// Copy copies the nodes and edges from the Graphviz AST source graph to the
-// destination graph. Edge direction is maintained if present.
-func Copy(dst Builder, src *ast.Graph) (err error) {
+// copyGraph copies the nodes and edges from the Graphviz AST source graph to
```
The documentation above should be changed to reflect this; delete "For DOT-files containing more than one graph, use dot.Copy."

mewmew replied:
That's right. Forgot to in all the excitement.

mewmew replied:
done

kortschak commented:
I don't know if `gocc` will let you do this, but can you make the generated code warning be in the form "// Code generated by <command>; DO NOT EDIT.", preferably on the first line, above the copyright notice and separated by a blank line.

This magic comment form is recognised by github (and other things) to avoid displaying generated code diffs unless requested.

mewmew commented:
> I don't know if gocc will let you do this, but can you make the generated code warning be in the form "// Code generated by ; DO NOT EDIT.", preferably on the first line, above the copyright notice and separated by a blank line.

Now the [gocc PR](https://github.com/goccmack/gocc/pull/46) has been sent and the formats/dot code has been updated in accordance with golang/go#13560.

kortschak commented on formats/dot/internal/errors/errors.go:
```diff
@@ -2,7 +2,7 @@
 // license that can be found in the LICENSE file.
 
-// generated by gocc; DO NOT EDIT.
+// Code generated by gocc; DO NOT EDIT.
```
It seems that github is stricter than the decision for this form. This diff should not be visible - see for example https://github.com/gonum/blas/commit/0285886d824b705f344567a692d06cf96179ad67. It probably needs to be on the first line.

mewmew replied:
Hmm.

It looks like it hides some of the files, even when a copyright header was preceding the "Code generated" comment. For instance, in my browser ` formats/dot/internal/lexer/transitiontable.go` is hidden, even though it contains:

```go
// Copyright ©2017 The gonum Authors. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

// Code generated by gocc; DO NOT EDIT.
```

https://github.com/gonum/graph/pull/198/files

Looking a bit closer, it seems that was because of the length of the file, as ` formats/dot/ast/ast.go` was also hidden by default, but this file is hand written and does not contain a "Code generated" comment.

mewmew replied:
During the PR process, I've manually added the copyright notice to auto-generated files. Not even sure if this makes sense, or if those can be copyrighted in the same way since they are produced by a tool and not us.

If we simply regenerate the files, the "Code generated by" comment should be on top, without a copyright header. This would solve this issue you highlighted, and also facilitate regeneration in the future. Another approach would be to automatically insert the copyright header between the "Code generated by" comment and the `package` keyword.

I would prefer not the first approach. What are your thoughts?

mewmew replied:
Off topic, but the content of blas/cgo/level1doubleBench_auto_test.go seems a bit weird. Is the comment supposed to contain `\n` characters?

From https://github.com/gonum/blas/commit/0285886d824b705f344567a692d06cf96179ad67#diff-c6db70f548cdc8ca223ae67f49e9d561
```go
// Code generated by \"go run github.com/gonum/blas/testblas/benchautogen/autogen_bench_level1double.go\"; DO NOT EDIT.\n
```


https://github.com/gonum/blas/commit/0285886d824b705f344567a692d06cf96179ad67#diff-c6db70f548cdc8ca223ae67f49e9d561

kortschak replied:
Large diffs get hidden as well. It's unfortunate gocc doesn't give greater flexibility (I've used ragel which does this kind of thing very simply). I will think about this.

Good spotting on the `\n`. I'll fix that. Thanks.

mewmew replied:
> t's unfortunate gocc doesn't give greater flexibility 

If you figure out a clean and generic approach we may add it to gocc. How did it work in ragel? The Gocc team is really lovely, and they often have a quick response time to new design ideas and PRs.

kortschak replied:
I think the copyright header does need to be on generated code. The generated code is the artifact, the process to generate it may be punching keys, or getting a machine to pretend to punch keys. The path to the product is not really relevant.

The way that ragel works is that the generated code is produced from a template ragel file. An example ragel file [here](https://github.com/cayleygraph/cayley/blob/master/quad/nquads/raw.rl) which generates [this](https://github.com/cayleygraph/cayley/blob/master/quad/nquads/raw.go). Although even that inserts a line directive - in a way that actually explicitly does not match the canonical form.

I guess the best approach is to add some code to the make scripts that will replace the warning with the warning suffixed with the copyright notice.

Sorry for all this rigmarole.

mewmew replied:
> I think the copyright header does need to be on generated code. The generated code is the artifact, the process to generate it may be punching keys, or getting a machine to pretend to punch keys. The path to the product is not really relevant.

Right.

> I guess the best approach is to add some code to the make scripts that will replace the warning with the warning suffixed with the copyright notice.

Would you be willing to take a look at writing such a script? I'm not sure entirely how to do it in a clean manner, and where it makes sense to hook it in.

Should it be something written in Go with go:generate, or perhaps a shell script (e.g. `sed`) with regular expressions?

> Sorry for all this rigmarole.

That's ok. I am very glad that you have done a thorough review and I've vastly appreciated the feedback! Thanks.

Truth to be told, I had hoped the generated code would not need a copyright notice, seeing as we did not really write it as such. But that's fine. Once this last remark is addressed we should be closer to merging the PR and get decoding support into gonum/graph, which really excites me.

Have a lovely day

Cheers /u

kortschak commented:
Here you go.

```diff
diff --git a/formats/dot/internal/Makefile b/formats/dot/internal/Makefile
index 196e70f..e31b461 100644
--- a/formats/dot/internal/Makefile
+++ b/formats/dot/internal/Makefile
@@ -1,6 +1,7 @@
 gen: dot.bnf
        gocc $<
        # TODO: Remove once https://github.com/goccmack/gocc/issues/36 gets resolved.
+       ./paste_copyright.bash
        find . -type f -name '*.go' | xargs goimports -w
 
 debug_lexer: dot.bnf
diff --git a/formats/dot/internal/paste_copyright.bash b/formats/dot/internal/paste_copyright.bash
new file mode 100755
index 0000000..eba0c7a
--- /dev/null
+++ b/formats/dot/internal/paste_copyright.bash
@@ -0,0 +1,17 @@
+#!/usr/bin/env bash
+
+find . -type f -name '*.go' \
+| xargs sed -i -e "s|// Code generated by gocc; DO NOT EDIT.|\
+// Code generated by gocc; DO NOT EDIT.\n\
+//\n\
+// This file is dual licensed under CC0 and The gonum license.\n\
+//\n\
+// Copyright ©2017 The gonum Authors. All rights reserved.\n\
+// Use of this source code is governed by a BSD-style\n\
+// license that can be found in the LICENSE file.\n\
+//\n\
+// Copyright ©2017 Robin Eklind.\n\
+// This file is made available under a Creative Commons CC0 1.0\n\
+// Universal Public Domain Dedication.\n\
+|"
+
```

vladimir-ch commented on encoding/dot/decode.go:
```diff
@@ -0,0 +1,237 @@
+	gen.subStart = append(gen.subStart, len(gen.subNodes))
+}
+
+// pushSubgraph pops the node start index of the active subgraph from the stack,
```
popSubgraph

mewmew replied:
Good spotting this.

mewmew commented:
> Here you go.

@kortschak Thanks a lot for the extensive review! The shell script has now been included in the regeneration process.

I believe all comments have been addressed, unless I missed something. Still kind of wished that you could have a checkbox to mark a comment as done in PRs.

In either case, when you both feel comfortable with it, feel free to merge this PR.

We can continue to polish the code in subsequent PRs. For instance, I would very much like to extend existing graph representations that have defined their own Node and Edge types to implement the encoding/dot.Builder interface by adding `NewNode` and `NewEdge` methods.

I could prepare PRs for those in the near future. May need to ask some questions regarding what data should be kept, and in which format, but one thing at the time.

Wish you both a lovely day : )

Cheers /u

mewmew commented:
> Do you want to make sure the hand coded files are also under CC0 (spray the license header over the things you wrote in formats)?

Done.

mewmew commented:
> Also, before I squash and merge, would you please send a PR to gonum/license adding yourself to A+C?

Will do.

mewmew commented:
> > Also, before I squash and merge, would you please send a PR to gonum/license adding yourself to A+C?
>
> Will do.

Oh, as it turns out, I was already added to the license repo at some point.

Feel free to squash and merge.

kortschak commented:
Thank you.

mewmew commented:
Yay! Thanks a lot for including this. It will make playing with gonum graphs even more pleasant.

mewmew commented:
> > Do you want to make sure the hand coded files are also under CC0 (spray the license header over the things you wrote in formats)?
>
> Done.

A closer look reveals that the updated comment was missing from a couple of files. I'll prepare a new PR to update those.
